### DISCUSSION TITLE
How complete is the API?

### DISCUSSION THREAD

**Discussion by dayunbao** [supriya-project/supriya#412] [Q&A]
Hi!  I've been learning SuperCollider for a project I have in mind.  I'm planning to build a hardware instrument.  It will be a modular groovebox.  The modules will be hot-swappable MIDI controllers that will have a groovebox-like form factor when all put together.  I want to use SC for sound generation.  However, I beginning to question whether sclang is suitable for what I want to do.   It wasn't designed to be an all-purpose programming language, even though it has a lot of the usual OOP stuff in it.  I've seen warnings from experienced SC users about sclang's limitations, and what a nightmare it can be to attempt a large software project with it.  I'm also a Python developer, and would _much_ rather work with Python than sclang.  So I was excited when I found Supriya. 
 However, before I switch over, I wanted to know how complete is the API?  Is everything sclang is capable of doing possible with Supriya?  I don't want to find out, weeks into the project, that some vital piece of functionality isn't available in Supriya.

Thanks!

**Comment by dayunbao** [comment]
I see that the API doesn't include SC's MIDI related classes (MIDIClient, MIDIIn,  MIDIOut, etc.). Was that because there are already Python MIDI libraries?  Would we be expected to use one of those with Supriya?  MIDI is essential to my project.

**Comment by josephine-wolf-oberholtzer** [comment]
Didn't see this because I apparently wasn't subscribed to my own project's discussions.

We've talked over email already, but the short story w/r/t MIDI is that there are already a number of Python libraries for handling MIDI and I didn't want to either a) bless one by making a wrapper for it or b) implement any MIDI logic from scratch.

The answer to the broader question is that no, supriya does not attempt to re-implement everything implemented in sclang. I try to keep supriya more narrowly scoped to server management, OSC, synthdefs, and timing-related concerns (clocks, basic patterns, etc.).

Supriya is intended as a foundational layer that other projects can be built on. I'm not currently planning on implementing (for example) an IDE, UI bindings, any of the graphical stuff from sclang, any of the micro-language / DSL stuff for live coding, etc. I think most of those topics either don't have obvious single solutions, will see limited use, or will generally be a maintenance burden on me.

I _am_ hoping to implement code for modeling simple DAW logic: mixers, channels/tracks, sends/receives, monitoring, etc.

**Comment by dayunbao** [comment]
Makes sense! It's already amazing to have everything that supriya offers.
